atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/ring/RingGraphTest.java

```diff
@@ -0,0 +1,108 @@
+
+    private static final Map<Integer, Integer> VALID_RING = Map.of(0, 2, 1, 0, 2, 1);
+
+    @Test
+    public void fromPartialGeneratesRandomRingWhenAllKeysEmpty() {
+        Map<Integer, Optional<Integer>> emptyEdges = Map.of(1, Optional.empty(), 2, Optional.empty());
+        assertThatCode(() -> RingGraph.fromPartial(emptyEdges)).doesNotThrowAnyException();
+    }
+
+    @Test
+    public void fromPartialThrowsWhenSomeEntriesAreEmpty() {
+        Map<Integer, Optional<Integer>> partiallyEmptyRing = Map.of(0, Optional.of(1), 1, Optional.empty());
+        assertThatLoggableExceptionThrownBy(() -> RingGraph.fromPartial(partiallyEmptyRing))
+                .isInstanceOf(SafeIllegalArgumentException.class)
+                .hasMessageContaining("Graph contains missing entries, thus cannot be made into a ring.")
+                .hasExactlyArgs(SafeArg.of("ring", partiallyEmptyRing));
+    }
+
+    @Test
+    public void fromThrowsMissingEntriesWhenNodeReferencedDoesNotExist() {
+        Map<Integer, Integer> missingNodesRing = Map.of(0, 2, 1, 0);
+        assertThatLoggableExceptionThrownBy(() -> RingGraph.from(missingNodesRing))
+                .isInstanceOf(RingValidationException.class)
+                .hasExactlyArgs(
+                        SafeArg.of("type", RingValidationException.Type.MISSING_ENTRIES),
+                        SafeArg.of("ring", missingNodesRing));
+    }
+
+    @Test
+    public void fromThrowsCycleExceptionWhenUnableToReachEveryNode() {
+        Map<Integer, Integer> incompleteRing = Map.of(0, 1, 1, 1, 2, 0);
+        assertThatLoggableExceptionThrownBy(() -> RingGraph.from(incompleteRing))
+                .isInstanceOf(RingValidationException.class)
+                .hasExactlyArgs(
+                        SafeArg.of("type", RingValidationException.Type.CYCLE), SafeArg.of("ring", incompleteRing));
+    }
+
+    @Test
+    public void fromThrowsCycleExceptionWhenRootNodeNotRevisited() {
+        Map<Integer, Integer> incompleteRing = Map.of(0, 1, 1, 1);
+        assertThatLoggableExceptionThrownBy(() -> RingGraph.from(incompleteRing))
+                .isInstanceOf(RingValidationException.class)
+                .hasExactlyArgs(
+                        SafeArg.of("type", RingValidationException.Type.CYCLE), SafeArg.of("ring", incompleteRing));
+    }
+
+    @Test
+    public void fromReturnsRandomlyGeneratedRingForEmptyRing() {
+        RingGraph ringGraph = RingGraph.from(VALID_RING);
+        boolean uniqueAtLeastOnce = false;
+        for (int iteration = 0; iteration < 10; iteration++) {
+            if (!ringGraph.generateNewRing().asMap().equals(ringGraph.asMap())) {
+                uniqueAtLeastOnce = true;
+                break;
+            }
+        }
+        assertThat(uniqueAtLeastOnce)
+                .as("Check ring generation multiple times as we could randomly generate an identical ring")
+                .isTrue();
+    }
+
+    @Test
+    public void generateNewRingReturnsValidRing() {
```

Comment:
repeatedGenerateNewRingDoesNotThrow?

Reply:
eh, I think the current test is more descriptive, as we're more concerned with generating an invalid ring, which then breaks later on, and _not_ for corruption reasons